File: SYP.Web/Modules/Email/EmailQueue/EmailQueueForm.cs
namespace SYP.Email.Forms;

[FormScript("Email.EmailQueue")]
[BasedOnRow(typeof(EmailQueueRow), CheckNames = true)]
public class EmailQueueForm
{
    [Tab("Alıcılar")]
    [TextAreaEditor(Rows = 3), Placeholder("Her satıra bir email adresi")]
    public string ToAddresses { get; set; }

    [TextAreaEditor(Rows = 2)]
    public string CcAddresses { get; set; }

    [TextAreaEditor(Rows = 2)]
    public string BccAddresses { get; set; }

    [Tab("İçerik")]
    public int SmtpSettingsId { get; set; }
    public int TemplateId { get; set; }
    public string Subject { get; set; }

    [HtmlContentEditor(Rows = 15)]
    public string Body { get; set; }

    [TextAreaEditor(Rows = 5)]
    public string BodyText { get; set; }

    [Tab("Gönderici")]
    [HalfWidth, EmailAddressEditor]
    public string FromAddress { get; set; }

    [HalfWidth]
    public string FromName { get; set; }

    [EmailAddressEditor]
    public string ReplyToAddress { get; set; }

    [Tab("Zamanlanmış Gönderim")]
    [HalfWidth]
    public int Priority { get; set; }

    [HalfWidth]
    public DateTime ScheduledAt { get; set; }

    [Tab("Durum")]
    [ReadOnly(true)]
    public int Status { get; set; }

    [ReadOnly(true)]
    public int RetryCount { get; set; }

    [ReadOnly(true)]
    public DateTime ProcessedAt { get; set; }

    [ReadOnly(true)]
    public DateTime SentAt { get; set; }

    [ReadOnly(true)]
    public DateTime NextRetryAt { get; set; }

    [TextAreaEditor(Rows = 5), ReadOnly(true)]
    public string ErrorMessage { get; set; }

    [Tab("Referans")]
    [HalfWidth]
    public string ReferenceType { get; set; }

    [HalfWidth]
    public string ReferenceId { get; set; }

    [TextAreaEditor(Rows = 5)]
    public string TemplateData { get; set; }
}

File: SYP.Web/Modules/ServerTypes/Email/EmailQueueForm.ts
import { DateEditor, EmailAddressEditor, EnumEditor, HtmlContentEditor, initFormType, IntegerEditor, LookupEditor, PrefixedContext, StringEditor, TextAreaEditor } from "@serenity-is/corelib";
import { EmailPriority } from "./EmailPriority";
import { EmailQueueStatus } from "./EmailQueueStatus";

export interface EmailQueueForm {
    ToAddresses: TextAreaEditor;
    CcAddresses: TextAreaEditor;
    BccAddresses: TextAreaEditor;
    SmtpSettingsId: LookupEditor;
    TemplateId: LookupEditor;
    Subject: StringEditor;
    Body: HtmlContentEditor;
    BodyText: TextAreaEditor;
    FromAddress: EmailAddressEditor;
    FromName: StringEditor;
    ReplyToAddress: EmailAddressEditor;
    Priority: EnumEditor;
    ScheduledAt: DateEditor;
    Status: EnumEditor;
    RetryCount: IntegerEditor;
    ProcessedAt: DateEditor;
    SentAt: DateEditor;
    NextRetryAt: DateEditor;
    ErrorMessage: TextAreaEditor;
    ReferenceType: StringEditor;
    ReferenceId: StringEditor;
    TemplateData: TextAreaEditor;
}

export class EmailQueueForm extends PrefixedContext {
    static readonly formKey = 'Email.EmailQueue';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!EmailQueueForm.init) {
            EmailQueueForm.init = true;

            var w0 = TextAreaEditor;
            var w1 = LookupEditor;
            var w2 = StringEditor;
            var w3 = HtmlContentEditor;
            var w4 = EmailAddressEditor;
            var w5 = EnumEditor;
            var w6 = DateEditor;
            var w7 = IntegerEditor;

            initFormType(EmailQueueForm, [
                'ToAddresses', w0,
                'CcAddresses', w0,
                'BccAddresses', w0,
                'SmtpSettingsId', w1,
                'TemplateId', w1,
                'Subject', w2,
                'Body', w3,
                'BodyText', w0,
                'FromAddress', w4,
                'FromName', w2,
                'ReplyToAddress', w4,
                'Priority', w5,
                'ScheduledAt', w6,
                'Status', w5,
                'RetryCount', w7,
                'ProcessedAt', w6,
                'SentAt', w6,
                'NextRetryAt', w6,
                'ErrorMessage', w0,
                'ReferenceType', w2,
                'ReferenceId', w2,
                'TemplateData', w0
            ]);
        }
    }
}

[EmailPriority, EmailQueueStatus]; // referenced types